### DISCUSSION TITLE
Does not support this configuration  ： redis sentinel

### DISCUSSION THREAD

**Discussion by dobixu** [apache/airflow#37556] [General] [answered]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

> [2023-11-14T11:31:46.663+0800] {base_events.py:1753} ERROR - Future exception was never retrieved
future: <Future finished exception=ValueError("'master_name' transport option must be specified.")>
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/kombu/transport/virtual/base.py", line 951, in create_channel
    return self._avail_channels.pop()
IndexError: pop from empty list
During handling of the above exception, another exception occurred: 
 Traceback (most recent call last):
  File "/usr/lib64/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.9/site-packages/celery/app/control.py", line 662, in enable_events
    return self.broadcast(
  File "/usr/local/lib/python3.9/site-packages/celery/app/control.py", line 776, in broadcast
    return self.mailbox(conn)._broadcast(
  File "/usr/local/lib/python3.9/site-packages/kombu/pidbox.py", line 330, in _broadcast
    chan = channel or self.connection.default_channel
  File "/usr/local/lib/python3.9/site-packages/kombu/connection.py", line 953, in default_channel
    self._ensure_connection(**conn_opts)
  File "/usr/local/lib/python3.9/site-packages/kombu/connection.py", line 459, in _ensure_connection
    return retry_over_time(
  File "/usr/local/lib/python3.9/site-packages/kombu/utils/functional.py", line 318, in retry_over_time
    return fun(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/kombu/connection.py", line 934, in _connection_factory
    self._connection = self._establish_connection()
  File "/usr/local/lib/python3.9/site-packages/kombu/connection.py", line 860, in _establish_connection
    conn = self.transport.establish_connection()
  File "/usr/local/lib/python3.9/site-packages/kombu/transport/virtual/base.py", line 975, in establish_connection
    self._avail_channels.append(self.create_channel(self))
  File "/usr/local/lib/python3.9/site-packages/kombu/transport/virtual/base.py", line 953, in create_channel
    channel = self.Channel(connection)
  File "/usr/local/lib/python3.9/site-packages/kombu/transport/redis.py", line 741, in __init__
    self.client.ping()
  File "/usr/local/lib/python3.9/site-packages/kombu/utils/objects.py", line 31, in __get__
    return super().__get__(instance, owner)
  File "/usr/lib64/python3.9/functools.py", line 993, in __get__
    val = self.func(instance)
  File "/usr/local/lib/python3.9/site-packages/kombu/transport/redis.py", line 1254, in client
    return self._create_client(asynchronous=True)
  File "/usr/local/lib/python3.9/site-packages/kombu/transport/redis.py", line 1210, in _create_client
    return self.Client(connection_pool=self.async_pool)
  File "/usr/local/lib/python3.9/site-packages/kombu/transport/redis.py", line 1248, in async_pool
    self._async_pool = self._get_pool(asynchronous=True)
  File "/usr/local/lib/python3.9/site-packages/kombu/transport/redis.py", line 1433, in _get_pool
    return self._sentinel_managed_pool(asynchronous)
  File "/usr/local/lib/python3.9/site-packages/kombu/transport/redis.py", line 1423, in _sentinel_managed_pool
    raise ValueError(
ValueError: 'master_name' transport option must be specified.

### What you think should happen instead

The correct configuration should support a Redis Sentinel cluster. Please provide a airflow.cnf default Redis Sentinel cluster configuration. Thank you very much.

## It works as expected on single Redis Configuration :
[celery]
broker_url = redis://:Passw0rd@172.16.0.202:16379/0

## Stably based on the  Redis Sentinel cluster runs normally,I have tried multiple configuration approaches, but none of them have resolved the issue. The configurations attempted are as follows:
1. Example
> [celery]
broker_url = redis+sentinel://:Passw0rd@172.16.0.202:26379,172.16.0.204:26379,172.16.0.206:26379/master

2. Example
> [celery]
broker_url = sentinel://Passw0rd@172.16.0.202:26379/0;sentinel://Passw0rd@172.16.0.204:26379/0;sentinel://Passw0rd@172.16.0.206:26379/master
BROKER_TRANSPORT_OPTIONS = {"master_name": "master"}

### Operating System

AlmaLinux release 9.2 

### Versions of Apache Airflow Providers

2.3.3/2.7.2

### Deployment

Docker-Compose

### Deployment details

> version: '3.8'
networks:
  airflow:
    ipam:
      config:
        - subnet: 172.178.0.0/24
x-airflow-common:
  &airflow-common
  image: registry.manyun-inc.com/services/airflow:2.7.2
  #image: registry.manyun-inc.com/services/airflow:2.3.3
  environment:
    &airflow-common-env
    TZ: 'Asia/Shanghai'
    AIRFLOW_HOME: '/data/airflow'
    AIRFLOW_CORE_DEFAULT_TIMEZONE: Asia/Shanghai 
    AIRFLOW_WEBSERVER_DEFAULT_TIMEZONE: Asia/Shanghai 
    #  env_file:
    #- mysql.env
  volumes:
          #    - /etc/localtime:/etc/localtime:ro
          #    - /etc/timezone:/etc/timezone:ro
    - /var/run/docker.sock:/var/run/docker.sock
    - ./dags:/data/airflow/dags
    - ./logs:/data/airflow/logs
      #- ./d_task:/data/airflow/d-task
    - ./plugins:/data/airflow/plugins
    - ./airflow.cfg:/data/airflow/airflow.cfg
services:
  airflow-webserver:
    <<: *airflow-common
    container_name: airflow.webserver
    command: webserver
    ports:
      - 38080:8080
    healthcheck:
      test: ["CMD", "curl", "--fail", "http://localhost:8080/health"]
      interval: 10s
      timeout: 10s
      retries: 5
    restart: always
    networks:
      - airflow
  airflow-scheduler1:
    <<: *airflow-common
    container_name: airflow.scheduler_1
    command: scheduler
    healthcheck:
      test: "ps axu | grep -vE 'grep|restart' | grep 'scheduler'"
      interval: 10s
      timeout: 10s
      retries: 5
    restart: always
    networks:
      - airflow
  airflow-worker:
    <<: *airflow-common
    container_name: airflow.worker
    command: celery worker
    healthcheck:
      test:
        - "CMD-SHELL"
        - 'celery --app airflow.executors.celery_executor.app inspect ping -d "celery@$${HOSTNAME}"'
      interval: 10s
      timeout: 10s
      retries: 5
    restart: always
    networks:
      - airflow
  airflow-init:
    <<: *airflow-common
    environment:
      <<: *airflow-common-env
    container_name: airflow.init
    command: db migrate
    #command: db init
    networks:
      - airflow
  flower:
    <<: *airflow-common
    container_name: airflow.flower
    command: celery flower
    ports:
      - 5555:5555
    healthcheck:
      test: ["CMD", "curl", "--fail", "http://localhost:5555/"]
      interval: 10s
      timeout: 10s
      retries: 5
    restart: always
    networks:
      - airflow

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by TicyYang** [comment, marked as answer]
Hi, 
I have also tried many times and found that the following configuration approach connects successfully:
```
[celery]
broker_url = sentinel://<redis_host_ip>:26379/0;sentinel://<redis_host_ip>:26380/0;sentinel://<redis_host_ip>:26381/0

result_backend = db+mysql://<user>:<password>@<mysql_host_ip>:3306/airflow_db


[celery_broker_transport_options]
master_name = mymaster
```

**Reply by dobixu** [reply]
Thanks you ,sir! Reids Sentinel is working!

**Reply by longpt233** [reply]
it work but just work with the first sentinel. if the first down, the airflow was crash (503). did you test this case @TicyYang

**Reply by TicyYang** [reply]
@longpt233 During my tests, when the first Sentinel is down, Airflow can still work.
However, if the Redis master fails (no matter which node the master is on), Airflow must wait for the failover to finish. During the failover, if you open the Flower Web UI, you can see that all worker statuses are offline.

It may not be related to the issue you are encountering, but may I ask if you are using Redis ACL?

**Reply by longpt233** [reply]
when I start flower and worker, I saw the log ```[2024-07-23 08:52:05,770: INFO/MainProcess] Connected to sentinel://172.21.0.5:26379/0```. It only the first sentinel, Do you have the same log line @TicyYang  ? my config: 
```
broker_url=sentinel://172.21.0.5:26379/0;sentinel://172.21.0.6:26379/0;sentinel://172.21.0.7:26379/0
```
docker-sentinel: 
![image](https://github.com/user-attachments/assets/eaa50dd1-8680-4412-9656-ab9204870d42)
flower when kill the fisrt sentinel: 
![image](https://github.com/user-attachments/assets/5d54d36a-bd6c-4117-95a4-7dd9eb0ca374)
![image](https://github.com/user-attachments/assets/6480fad9-1827-4a04-a6ce-5f58f7001389)

**Reply by TicyYang** [reply]
@longpt233  The architecture I previously tested consisted of three VMs, so I spent some time testing with Docker.
I have the same log `Connected to sentinel://...` when I started the worker and the flower. I think this is correct, and I believe that the IP set for the `broker_url` should be the host's IP, not the container's.
Even if I stop the first sentinel, the worker and flower are not affected and are able to execute tasks.
Somehow, although the worker, Redis, and sentinel work fine, the "Broker" tab on the Flower Web UI always shows "Error 500".

**Reply by longpt233** [reply]
thank you. in concise,  I think 500 just a litter bug. everything run ok without a sentinel

**Reply by TicyYang** [reply]
I believe the problem lies with the IP because when I set `broker_url` to the container's IP, such as: `sentinel://172.29.0.4:26379/0;sentinel://172.29.0.5:26379/0;sentinel://172.29.0.7:26379/0`
If I stop the first sentinel, tasks cannot be executed, and the flower logs shows: 
![螢幕擷取畫面 2024-07-26 003713](https://github.com/user-attachments/assets/3c69fae0-d515-489f-a5df-3b926bfe7752)
![螢幕擷取畫面 2024-07-26 005714](https://github.com/user-attachments/assets/2c288eee-401b-4363-b8e5-f7078b40ab48)
And if I set it to host's IP, nothing happens.
